apps/lchfn.1
.\" A man page for lchfn
.\" Copyright (C) 2004 Red Hat, Inc.
.\"
.\" This is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU Library General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU Library General Public
.\" License along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" Author: Miloslav Trmac <mitr@redhat.com>
.TH lchfn 1 "Sep 20 2004" libuser

.SH NAME
lchfn \- Change finger information

.SH SYNOPSIS
lchfn [\fIOPTION\fR]... [\fIuser\fR]

.SH DESCRIPTION
Displays and allows changing information about \fIuser\fR 
that is available using the
.BR finger (1)
command (usually stored in the "gecos" field of \fI/etc/passwd\fR).

If the \fIuser\fR argument is not provided,
username of the invoking user is used;
the \fIuser\fR argument is ignored if
.B lchfn
is run set-uid to a different user.

Entering an empty string (by pressing Enter) at a prompt
is interpreted as accepting a default value;
to erase an attribute (or to use an empty string as the attribute value),
enter a single dot character.

.SH OPTIONS
.TP
\fB\-i\fR, \fB\-\-interactive\fR 
Ask all questions when connecting to the user database,
even if default answers are set up in
.B libuser
configuration.

.SH EXIT STATUS
The exit status is 0 on success, 1 on error.


NEWS
0.53
* Support UID and GID values larger than LONG_MAX (#124967)
* FIXME lusermod group renaming fix
* Miscellaneous bug and memory leak fixes

0.52.6
* Mark more strings for translation
* Make error reporting more consistent and more verbose, output error messages
  on stderr.
* Port sasldb backend to Cyrus SASL v2, make it at least minimally usable

0.52.5
* Fix home directory renaming in ADMIN.modifyUser (#135280)
* Further Python reference counting fixes

0.52.4
* Memory leak fixes

0.52.3
* Fix compilation without libuser headers already installed (#134085)

0.52.2
* Allow LDAP connection using ldaps, custom ports or without TLS (original
  patch from Pawel Salek).

0.52.1
* Fix freecon() of uninitialized value in files/shadow module

0.52
* Usable LDAP backend
* Miscellaneous bug fixes

0.51.12
* Don't claim success and exception at the same time (#133479)
* LDAP fixes, second round
* Various other bug fixes

0.51.11
* Allow documented optional arguments in Python
  ADMIN.{addUser,modifyUser,deleteUser} (#119812)
* Add man pages for lchfn and lchsh
* LDAP fixes, first round
* Avoid file conflict on multilib systems
* Call ldconfig correctly

0.51.10
* Don't attempt to lookup using original entity name after entity
  modification (rename in particular) (#78376, #121252)
* Fix copying of symlinks from /etc/skel (#87572, original patch from gLaNDix)
* Make --enable-quota work, and fix the quota code to at least compile (#89114)
* Fix several bugs (#120168, original patch from Steve Grubb)
* Don't hardcode python version in spec file (#130952, from Robert Scheck)
* Properly integrate the SELinux patch, it should actually be used now, even
  though it was "enabled" since 0.51.7-6

0.51.9
* Fix various typos
* Document library interfaces
* Build all shared libraries with -fPIC (#72536)

0.51.8
* Update to build with latest autotools and gtk-doc
* Update ALL_LINGUAS and POTFILES.in
* Rebuild to depend on newer openldap

0.51.7-7
* fix is_selinux_enabled call

0.51.7-3
* Add SELinux support

0.51.7
* ldap: set error codes correctly when we encounter failures initializing
* don't double-close modules which fail initialization
* ldap: don't set an error in cases where one is already set

0.51.6
* use a crypt salt consistent with the defaults/crypt_style setting when
  setting new passwords (#79337)

0.51.5
* expose lu_get_first_unused_id() as a package-private function
* provide libuser.ADMIN.getFirstUnusedUid and libuser.ADMIN.getFirstUnusedGid
  in python

0.51.4
* fix not freeing resources properly in files.c(generic_is_locked), spotted by
  Zou Pengcheng

0.51.2
* degrade gracefully
* build with --with-pic and -fPIC
* remove unpackaged man page

0.51.1-2
* translation updates

0.51.1-1
* doc updates -- cvs tree moved
* language updates
* disallow weird characters in account names

* automated rebuild

0.51
* files: ignore blank lines in files
* libuser: disallow creation of accounts with names containing whitespace,
  control characters, or non-ASCII characters

0.50.2
* refresh translations
* fix a heap-corruption bug in the python bindings

0.50
* bump version
* refresh translations

0.49.102
* ldap: cache an entity's dn in the entity structure to try to speed things up

0.49.101-2
* add missing buildreqs on cyrus-sasl-devel and openldap-devel (#59456)
* translation refresh

0.49.101-1
* fix python bindings of enumerateFull functions
* adjust prompter wrapping to not error out on successful returns

0.49.100
* be more careful about printing error messages
* fix refreshing after adding of accounts
* ldap: try to use a search to convert names to DNs, and only fall back to
  guessing if it turns up nothing
* files: fix an off-by-one in removal of entries

0.49.99
* refresh translations
* fix admin() constructor comments in the python module

0.49.98
* automatically refresh entities after add, modify, setpass, removepass,
  lock, and unlock operations
* remove debug spewage when creating and removing mail spools
* files: fix saving of multi-valued attributes
* rename MEMBERUID attribute for groups to MEMBERNAME

0.49.97
* files: fix bug in removals
* ldap: revert attempts at being smart at startup time, because it makes UIs
  very messy (up the three whole dialogs just to start the ldap stuff!)

0.49.96
* fix thinko in dispatch routines

0.49.95
* lgroupmod: fix thinko

0.49.93
* move shadow initialization for groups to the proper callback
* rework locking in the files module to not require that files be writable

* expose lu_strerror()
* add various typedefs for types used by the library

0.49.92
* add removepass() functions

* lchfn,lchsh,lpasswd - reorder PAM authentication calls
* include API docs in the package

0.49.91
* ldap: finish port to new API
* sasl: finish port to new API (needs test)
* libuser: don't commit object changes before passing data to service
  functions which might need differing data sets to figure out what to
  change (for example, ldap)

0.49.90
* bind the internal mail spool creation/removal functions for python

* renamed the python module
* revamped internals to use gobject's gvalues and gvaluearrays instead of
  glists of cached strings
* add enumeration-with-data functions to the C library

* require linuxdoc-tools instead of sgml-tools for rawhide

* fixup build files to allow building for arbitrary versions of python

0.32
* link the python module against libpam
* attempt to import the python modules at build-time to verify dependencies

0.31
* fix a file-parsing bug that popped up in 0.29's mmap modifications

0.30
* quotaq: fix argument order when reading quota information
* user_quota: set quota grace periods correctly
* luseradd: never create home directories for system accounts

* add da translation files
* update translations

0.29
* add an explicit build dependency on jade (for the docs)

* HUP nscd on modifications
* userutil.c: mmap files we're reading for probable speed gain
* userutil.c: be conservative with the amount of random data we read

apps/lchfn.c
/*
 * Copyright (C) 2001,2002 Red Hat, Inc.
 *
 * This is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#ident "$Id$"

#ifdef HAVE_CONFIG_H
#include "../config.h"
#endif
#include <sys/stat.h>
#include <libintl.h>
#include <locale.h>
#include <popt.h>
#include <pwd.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "../lib/user.h"
#include "apputil.h"

#define NAME_KEY		"lchfn/name"
#define SURNAME_KEY		"lchfn/surname"
#define GIVENNAME_KEY		"lchfn/givenname"
#define OFFICE_KEY		"lchfn/office"
#define OFFICEPHONE_KEY		"lchfn/officephone"
#define HOMEPHONE_KEY		"lchfn/homephone"
#define EMAIL_KEY		"lchfn/email"

int
main(int argc, const char **argv)
{
	const char *user = NULL, *gecos = NULL, *sn, *gn, *email, *filtered;
	const char *name, *office, *officephone, *homephone;
	struct lu_context *ctx = NULL;
	struct lu_error *error = NULL;
	struct lu_ent *ent = NULL;
	GValueArray *values;
	GValue *value, val;
	int interactive = FALSE;
	int c;
	struct lu_prompt prompts[7];
	poptContext popt;
	struct poptOption options[] = {
		{"interactive", 'i', POPT_ARG_NONE, &interactive, 0,
		 "prompt for all information", NULL},
		POPT_AUTOHELP POPT_TABLEEND
	};
	char **fields;
	size_t fields_len;
	size_t pcount, i;
	struct passwd *pwd = NULL;

	/* Set up i18n. */
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	setlocale(LC_ALL, "");

	/* Parse command-line arguments. */
	popt = poptGetContext("lchfn", argc, argv, options, 0);
	poptSetOtherOptionHelp(popt, _("[OPTION...] [user]"));
	c = poptGetNextOpt(popt);
	if (c != -1) {
		fprintf(stderr, _("Error parsing arguments: %s.\n"),
			poptStrerror(c));
		poptPrintUsage(popt, stderr, 0);
		exit(1);
	}
	user = poptGetArg(popt);

	/* If no user name was specified, or we're running in a setuid
	 * environment, force the user name to be the current user. */
	if ((user == NULL) || (geteuid() != getuid())) {
		pwd = getpwuid(getuid());
		if (pwd != NULL) {
			user = g_strdup(pwd->pw_name);
		} else {
			fprintf(stderr, _("No user name specified, no name "
				"for uid %d.\n"), getuid());
			poptPrintUsage(popt, stderr, 0);
			exit(1);
		}
	}

	/* Authenticate the user to the "chfn" service. */
	lu_authenticate_unprivileged(user, "chfn");

	/* Give the user some idea of what's going on. */
	g_print(_("Changing finger information for %s.\n"), user);

	/* Start up the library. */
	ctx = lu_start(user, lu_user, NULL, NULL,
		       interactive ? lu_prompt_console :
		       lu_prompt_console_quiet, NULL, &error);
	if (ctx == NULL) {
		if (error != NULL) {
			fprintf(stderr, _("Error initializing %s: %s.\n"),
				PACKAGE,
				error ? error->string : _("unknown error"));
		} else {
			fprintf(stderr, _("Error initializing %s.\n"),
				PACKAGE);
		}
		return 1;
	}

	/* Look up the user's information. */
	ent = lu_ent_new();
	if (lu_user_lookup_name(ctx, user, ent, &error) == FALSE) {
		g_print(_("User %s does not exist.\n"), user);
	}

	/* Read the user's GECOS information. */
	values = lu_ent_get(ent, LU_GECOS);
	if (values != NULL) {
		value = g_value_array_get_nth(values, 0);
		if (G_VALUE_HOLDS_STRING(value)) {
			gecos = g_value_get_string(value);
		} else
		if (G_VALUE_HOLDS_LONG(value)) {
			gecos = g_strdup_printf("%ld", g_value_get_long(value));
			/* leak */
		} else {
			g_assert_not_reached();
		}
	} else {
		gecos = "";
	}

	/* Split the gecos into the prompt structures. */
	fields = g_strsplit(gecos, ",", G_N_ELEMENTS(prompts));

	/* Count the number of fields we got. */
	fields_len = 0;
	if (fields != NULL) {
		while (fields[fields_len] != NULL) {
			fields_len++;
		}
	}

	/* Fill out the prompt structures. */
	memset(prompts, 0, sizeof(prompts));
	pcount = 0;

	/* The first prompt is for the full name. */
	name = (fields_len > 0) ? fields[0] : NULL;
	prompts[pcount].key = NAME_KEY;
	prompts[pcount].prompt = N_("Full Name");
	prompts[pcount].domain = PACKAGE;
	prompts[pcount].visible = TRUE;
	prompts[pcount].default_value = name;
	pcount++;

	/* If we have it, prompt for the user's surname. */
	values = lu_ent_get(ent, LU_SN);
	if (values != NULL) {
		value = g_value_array_get_nth(values, 0);
		sn = NULL;
		if (G_VALUE_HOLDS_STRING(value)) {
			sn = g_value_get_string(value);
		} else
		if (G_VALUE_HOLDS_LONG(value)) {
			sn = g_strdup_printf("%ld", g_value_get_long(value));
			/* leak */
		} else {
			g_assert_not_reached();
		}
		prompts[pcount].key = SURNAME_KEY;
		prompts[pcount].prompt = N_("Surname");
		prompts[pcount].domain = PACKAGE;
		prompts[pcount].visible = TRUE;
		prompts[pcount].default_value = sn;
		pcount++;
	}

	/* If we have it, prompt for the user's givenname. */
	values = lu_ent_get(ent, LU_GIVENNAME);
	if (values != NULL) {
		gn = NULL;
		value = g_value_array_get_nth(values, 0);
		if (G_VALUE_HOLDS_STRING(value)) {
			gn = g_value_get_string(value);
		} else
		if (G_VALUE_HOLDS_LONG(value)) {
			gn = g_strdup_printf("%ld", g_value_get_long(value));
			/* leak */
		} else {
			g_assert_not_reached();
		}
		prompts[pcount].key = GIVENNAME_KEY;
		prompts[pcount].prompt = N_("Given Name");
		prompts[pcount].domain = PACKAGE;
		prompts[pcount].visible = TRUE;
		prompts[pcount].default_value = gn;
		pcount++;
	}

	/* Prompt for the user's office number. */
	office = (fields_len > 1) ? fields[1] : NULL;
	prompts[pcount].key = OFFICE_KEY;
	prompts[pcount].prompt = N_("Office");
	prompts[pcount].domain = PACKAGE;
	prompts[pcount].visible = TRUE;
	prompts[pcount].default_value = office;
	pcount++;

	/* Prompt for the user's office telephone number. */
	officephone = (fields_len > 2) ? fields[2] : NULL;
	prompts[pcount].key = OFFICEPHONE_KEY;
	prompts[pcount].prompt = N_("Office Phone");
	prompts[pcount].domain = PACKAGE;
	prompts[pcount].visible = TRUE;
	prompts[pcount].default_value = officephone;
	pcount++;

	/* Prompt for the user's home telephone number. */
	homephone = (fields_len > 3) ? fields[3] : NULL;
	prompts[pcount].key = HOMEPHONE_KEY;
	prompts[pcount].prompt = N_("Home Phone");
	prompts[pcount].domain = PACKAGE;
	prompts[pcount].visible = TRUE;
	prompts[pcount].default_value = homephone;
	pcount++;

	/* If we have it, prompt for the user's email. */
	values = lu_ent_get(ent, LU_EMAIL);
	if (values != NULL) {
		value = g_value_array_get_nth(values, 0);
		email = NULL;
		if (G_VALUE_HOLDS_STRING(value)) {
			email = g_value_get_string(value);
		} else
		if (G_VALUE_HOLDS_LONG(value)) {
			email = g_strdup_printf("%ld", g_value_get_long(value));
			/* leak */
		} else {
			g_assert_not_reached();
		}
		prompts[pcount].key = EMAIL_KEY;
		prompts[pcount].prompt = N_("E-Mail Address");
		prompts[pcount].domain = PACKAGE;
		prompts[pcount].visible = TRUE;
		prompts[pcount].default_value = email;
		pcount++;
	}

	/* Sanity check. */
	g_assert(pcount <= G_N_ELEMENTS(prompts));

	/* Ask the user for new values. */
	if (lu_prompt_console(prompts, pcount, NULL, &error) == FALSE) {
		g_print(_("Finger information not changed:  input error.\n"));
		exit(1);
	}

	/* Initialize the temporary value variable. */
	memset(&val, 0, sizeof(val));
	g_value_init(&val, G_TYPE_STRING);

	/* Now iterate over the answers and figure things out. */
	for (i = 0; i < pcount; i++) {
		if (prompts[i].value == NULL) {
			filtered = "";
		} else
		if (strcmp(prompts[i].value, ".") == 0) {
			filtered = "";
		} else {
			filtered = prompts[i].value;
		}
		g_value_set_string(&val, filtered);

		if (strcmp(prompts[i].key, NAME_KEY) == 0) {
			name = filtered;
			lu_ent_clear(ent, LU_COMMONNAME);
			if (strlen(filtered) > 0) {
				lu_ent_add(ent, LU_COMMONNAME, &val);
			}
		}

		if (strcmp(prompts[i].key, SURNAME_KEY) == 0) {
			sn = filtered;
			lu_ent_clear(ent, LU_SN);
			if (strlen(filtered) > 0) {
				lu_ent_add(ent, LU_SN, &val);
			}
		}

		if (strcmp(prompts[i].key, GIVENNAME_KEY) == 0) {
			gn = filtered;
			lu_ent_clear(ent, LU_GIVENNAME);
			if (strlen(filtered) > 0) {
				lu_ent_add(ent, LU_GIVENNAME, &val);
			}
		}

		if (strcmp(prompts[i].key, OFFICE_KEY) == 0) {
			office = filtered;
			lu_ent_clear(ent, LU_ROOMNUMBER);
			if (strlen(filtered) > 0) {
				lu_ent_add(ent, LU_ROOMNUMBER, &val);
			}
		}

		if (strcmp(prompts[i].key, OFFICEPHONE_KEY) == 0) {
			officephone = filtered;
			lu_ent_clear(ent, LU_TELEPHONENUMBER);
			if (strlen(filtered) > 0) {
				lu_ent_add(ent, LU_TELEPHONENUMBER, &val);
			}
		}

		if (strcmp(prompts[i].key, HOMEPHONE_KEY) == 0) {
			homephone = filtered;
			lu_ent_clear(ent, LU_HOMEPHONE);
			if (strlen(filtered) > 0) {
				lu_ent_add(ent, LU_HOMEPHONE, &val);
			}
		}

		if (strcmp(prompts[i].key, EMAIL_KEY) == 0) {
			email = filtered;
			lu_ent_clear(ent, LU_EMAIL);
			if (strlen(filtered) > 0) {
				lu_ent_add(ent, LU_EMAIL, &val);
			}
		}
	}

	g_value_reset(&val);

	/* Build a new gecos string. */
	gecos = g_strjoin(",",
			  name ?: "",
			  office ?: "",
			  officephone ?: "",
			  homephone ?: "",
			  NULL);

	/* Now we can free the answers. */
	for (i = 0; i < pcount; i++) {
		if (prompts[i].value != NULL) {
			if (prompts[i].free_value != NULL) {
				prompts[i].free_value(prompts[i].value);
				prompts[i].value = NULL;
			}
		}
	}

	/* Set the GECOS attribute. */
	g_value_set_string(&val, gecos);
	lu_ent_clear(ent, LU_GECOS);
	lu_ent_add(ent, LU_GECOS, &val);
	g_value_reset(&val);

	/* Try to save our changes. */
	if (lu_user_modify(ctx, ent, &error)) {
		g_print(_("Finger information changed.\n"));
		lu_hup_nscd();
	} else {
		g_print(_("Finger information not changed: %s.\n"),
			error ? error->string : _("unknown error"));
	}

	g_value_unset(&val);

	g_strfreev(fields);

	lu_ent_free(ent);

	lu_end(ctx);

	return 0;
}
